GitHubBranches.ps1

```diff
@@ -107,3 +107,514 @@ function Get-GitHubRepositoryBranch
+
+        [switch] $NoStatus
+    )
+
+    Write-InvocationLog
+
+    $elements = Resolve-RepositoryElements
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        'OwnerName' = (Get-PiiSafeString -PlainText $OwnerName)
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    $params = @{
+        UriFragment = "repos/$OwnerName/$RepositoryName/branches/$Name/protection"
+        Description =  "Getting branch protection status for $RepositoryName"
+        Method = 'Get'
+        AcceptHeader = $script:lukeCageAcceptHeader
+        AccessToken = $AccessToken
+        TelemetryEventName = $MyInvocation.MyCommand.Name
+        TelemetryProperties = $telemetryProperties
+        NoStatus = (Resolve-ParameterWithDefaultConfigurationValue -Name NoStatus -ConfigValueName DefaultNoStatus)
+    }
+
+    return Invoke-GHRestMethod @params
+}
+
+function Set-GitHubRepositoryBranchProtectionRule
+{
+<#
+    .SYNOPSIS
+        Set branch protection rules for a given GitHub repository.
+
+    .DESCRIPTION
+        Set branch protection rules for a given GitHub repository.
+
+        The Git repo for this module can be found here: http://aka.ms/PowerShellForGitHub
+
+    .PARAMETER Name
+        Name of the specific branch to be retrieved.  If not supplied, all branches will be retrieved.
+
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER OwnerName
+        Owner of the repository.
+        If not supplied here, the DefaultOwnerName configuration property value will be used.
+
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER StatusChecks
+        The list of status checks to require in order to merge into the branch.
+
+    .PARAMETER RequireUpToDateBranches
+        Require branches to be up to date before merging.
+
+    .PARAMETER EnforceAdmins
+        Enforce all configured restrictions for administrators.
+
+    .PARAMETER DismissalUsers
+        Specify which users can dismiss pull request reviews.
+
+    .PARAMETER DismissalTeams
+        Specify which teams can dismiss pull request reviews.
+
+    .PARAMETER DismissStaleReviews
+        If specified, approving reviews when someone pushes a new commit are automatically
+        dismissed.
+
+    .PARAMETER RequireCodeOwnerReviews
+        Blocks merging pull requests until code owners review them.
+
+    .PARAMETER RequiredApprovingReviewCount
+        Specify the number of reviewers required to approve pull requests. Use a number between 1
+        and 6.
+
+    .PARAMETER RestrictPushUsers
+        Specify which users have push access.
+
+    .PARAMETER RestrictPushTeams
+        Specify which teams have push access.
```

Comment:
Actually, now I see that you're explicitly looking up the slugs inside the function and are expecting the team names here.  But that's unnecessary work if the user happened to pipe in team objects (if you choose to support that).

Reply:
The Microsoft org has 2100 teams.

* Getting all of them with `Get-GitHubTeam -OrganizationName microsoft -NoStatus` adds an additional **22 seconds** to method execution.
* Calling it _with_ status adds an additional **104** seconds.
* Allowing users to call into the method providing the slugs for the teams would add an additional **0 seconds**. 

**Double** that time if they happen to pass in both Dismissal and Restriction teams, since you do it _twice_.  (By the way, you should also optimize the code to only call `Get-GitHubTeam` _once_ if any sort of team is provided).

I agree that exposing the existence of slugs makes the signature a bit uglier, however _not_ exposing the option of using slugs means that the API call can needlessly take additional _minutes_ to complete.

Reply:
And actually, saying that and then reviewing the code, it occurs to me that you don't have `OrganizationName` as a parameter, which means that users can't specify team names that are owned by an organization, which could be a problem if the repo that the rule is being applied to is one owned by an organization.